File: app/src/main/java/com/stevenfrew/beatprompter/cache/CachedCloudFile.java
package com.stevenfrew.beatprompter.cache;

import com.stevenfrew.beatprompter.cloud.CloudDownloadResult;
import com.stevenfrew.beatprompter.cloud.CloudFileInfo;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

import java.io.File;
import java.util.ArrayList;
import java.util.Date;

public abstract class CachedCloudFile extends CloudFileInfo {

    public File mFile;

    private final static String CACHED_FILE_PATH_ATTRIBUTE_NAME="path";

    CachedCloudFile(File file, String id, String name, Date lastModified,String subfolder)
    {
        super(id,name,lastModified,subfolder);
        mFile=file;
    }

    CachedCloudFile(File file, CloudFileInfo cloudFileInfo)
    {
        super(cloudFileInfo.mID,cloudFileInfo.mName,cloudFileInfo.mLastModified,cloudFileInfo.mSubfolder);
        mFile=file;
    }

    CachedCloudFile(CloudDownloadResult result)
    {
        this(result.mDownloadedFile,result.mCloudFileInfo);
    }

    CachedCloudFile(Element element)
    {
        super(element);
        mFile=new File(element.getAttribute(CACHED_FILE_PATH_ATTRIBUTE_NAME));
    }

    public static String getTokenValue(String line,int lineNumber, String... tokens)
    {
        ArrayList<String> values=getTokenValues(line, lineNumber, tokens);
        if(values.size()==0)
            return null;
        return values.get(values.size() - 1);
    }

    static ArrayList<String> getTokenValues(String line,int lineNumber, String... tokens)
    {
        ArrayList<Tag> tagsOut=new ArrayList<>();
        ArrayList<String> values=new ArrayList<>();
        if(!line.trim().startsWith("#")) {
            Tag.extractTags(line, lineNumber, tagsOut);
            for (Tag tag : tagsOut)
                for (String token : tokens)
                    if (tag.mName.equals(token))
                        if (tag.mValue != null)
                            values.add(tag.mValue.trim());
        }
        return values;
    }

    static boolean containsToken(String line,int lineNumber,String tokenToFind)
    {
        ArrayList<Tag> tagsOut=new ArrayList<>();
        if(!line.trim().startsWith("#")) {
            Tag.extractTags(line, lineNumber, tagsOut);
            for (Tag tag : tagsOut)
                if (tag.mName.equals(tokenToFind))
                    return true;
        }
        return false;
    }

    public abstract void writeToXML(Document d,Element element);

    public void writeToXML(Element element)
    {
        super.writeToXML(element);
        element.setAttribute(CACHED_FILE_PATH_ATTRIBUTE_NAME,mFile.getAbsolutePath());
    }

    public static CachedCloudFile createCachedCloudFile(CloudDownloadResult result)
    {
        try
        {
            return new AudioFile(result);
        }
        catch(InvalidBeatPrompterFileException ioe)
        {
            try {
                return new ImageFile(result);
            }
            catch(InvalidBeatPrompterFileException ibpfe1)
            {
                try {
                    return new MIDIAliasFile(result);
                }
                catch(InvalidBeatPrompterFileException ibpfe2)
                {
                    try {
                        return new SongFile(result);
                    }
                    catch(InvalidBeatPrompterFileException ibpfe3)
                    {
                        try {
                            return new SetListFile(result);
                        }
                        catch(InvalidBeatPrompterFileException ibpfe4)
                        {
                            // Not any kind of file we're interested in?
                            // We don't want this file.
                        }
                    }
                }
            }
        }
        return null;
    }
}

File: app/src/main/java/com/stevenfrew/beatprompter/cache/CachedCloudFile.kt
package com.stevenfrew.beatprompter.cache

import com.stevenfrew.beatprompter.cloud.CloudDownloadResult
import com.stevenfrew.beatprompter.cloud.CloudFileInfo
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.io.File
import java.util.*

abstract class CachedCloudFile : CloudFileInfo {
    @JvmField var mFile: File

    internal constructor(file: File, id: String, name: String, lastModified: Date, subfolder: String):super(id,name,lastModified,subfolder)
    {
        mFile=file;
    }

    internal constructor(file: File, cloudFileInfo: CloudFileInfo) : this(file,cloudFileInfo.mID, cloudFileInfo.mName, cloudFileInfo.mLastModified, cloudFileInfo.mSubfolder)

    internal constructor(result: CloudDownloadResult) : this(result.mDownloadedFile, result.mCloudFileInfo)

    internal constructor(element: Element) : super(element) {
        mFile = File(element.getAttribute(CACHED_FILE_PATH_ATTRIBUTE_NAME))
    }

    abstract fun writeToXML(d: Document, element: Element)

    override fun writeToXML(element: Element) {
        super.writeToXML(element)
        element.setAttribute(CACHED_FILE_PATH_ATTRIBUTE_NAME, mFile.absolutePath)
    }

    companion object {

        private const val CACHED_FILE_PATH_ATTRIBUTE_NAME = "path"

        fun getTokenValue(line: String, lineNumber: Int, vararg tokens: String): String? {
            val values = getTokenValues(line, lineNumber, *tokens)
            return if (values.size == 0) null else values[values.size - 1]
        }

        internal fun getTokenValues(line: String, lineNumber: Int, vararg tokens: String): ArrayList<String> {
            val tagsOut = ArrayList<Tag>()
            val values = ArrayList<String>()
            if (!line.trim { it <= ' ' }.startsWith("#")) {
                Tag.extractTags(line, lineNumber, tagsOut)
                for (tag in tagsOut)
                    for (token in tokens)
                        if (tag.mName == token)
                            values.add(tag.mValue.trim { it <= ' ' })
            }
            return values
        }

        internal fun containsToken(line: String, lineNumber: Int, tokenToFind: String): Boolean {
            val tagsOut = ArrayList<Tag>()
            if (!line.trim { it <= ' ' }.startsWith("#")) {
                Tag.extractTags(line, lineNumber, tagsOut)
                for (tag in tagsOut)
                    if (tag.mName == tokenToFind)
                        return true
            }
            return false
        }

        fun createCachedCloudFile(result: CloudDownloadResult): CachedCloudFile? {
            try {
                return AudioFile(result)
            } catch (ioe: InvalidBeatPrompterFileException) {
                try {
                    return ImageFile(result)
                } catch (ibpfe1: InvalidBeatPrompterFileException) {
                    try {
                        return MIDIAliasFile(result)
                    } catch (ibpfe2: InvalidBeatPrompterFileException) {
                        try {
                            return SongFile(result)
                        } catch (ibpfe3: InvalidBeatPrompterFileException) {
                            try {
                                return SetListFile(result)
                            } catch (ibpfe4: InvalidBeatPrompterFileException) {
                                // Not any kind of file we're interested in?
                                // We don't want this file.
                            }
                        }
                    }
                }
            }
            return null
        }
    }
}